RUFAS/input_manager.py

```diff
@@ -53,144 +77,368 @@ def _load_metadata(self, metadata_path: str = "input/example_metadata.json") ->
+                return False
+        else:
+            var_name = element_hierarchy[-1]
+            try:
+                input_data_value = reduce(lambda d, key: d[key], element_hierarchy, input_data)
+            except KeyError:
+                raise KeyError(f"Key {var_name} not found in input data")
 
-        # TODO in validate_element fun branch
-        # where element is found to be invalid, place this warning:
-        # om.add_warning("Invalid data", f"Invalid data found: {key=}; {value=}", info_map)
-        pass
+            type_validation_dict = {"string": self._validate_string_type_element,
+                                    "number": self._validate_num_type_element,
+                                    "array": self._validate_array_type_element,
+                                    "bool": self._validate_bool_type_element, }
+
+            try:
+                validity_checker = type_validation_dict[var_type]
+                is_valid = validity_checker(variable_properties, var_name, input_data_value)
+            except KeyError:
+                raise KeyError(f"Invalid type {var_type}: Element must be type number, array, string, or bool")
+
+            counter_dict["total_elements"] += 1
+            if is_valid:
+                counter_dict["valid_elements"] += 1
+                return True
+            else:
+                is_fixed = self._fix_data(variable_properties, element_hierarchy, input_data)
+                if is_fixed:
+                    counter_dict["fixed_elements"] += 1
+                else:
+                    counter_dict["invalid_elements"] += 1
+                return is_fixed
 
-    def _fix_data(self, key: str, value: Any) -> bool:
+    def _validate_array_type_element(self, variable_properties: Dict[str, Any], var_name: str,
+                                     input_data_value: list) -> bool:
+        """Validates an input data element of type array."""
+        info_map = {"class": self.__class__.__name__,
+                    "function": self._validate_array_type_element.__name__,
+                    }
+        maximum_length = variable_properties.get("maximum_length")
+        minimum_length = variable_properties.get("minimum_length")
+        is_in_range = True
+        if minimum_length is not None:
+            is_in_range = variable_properties["minimum_length"] <= len(input_data_value)
+            if not is_in_range:
+                warning_string = f"Array length less than {minimum_length}."
+                om.add_warning(warning_string, f"{var_name=}", info_map)
+                return is_in_range
+        if maximum_length is not None:
+            is_in_range = len(input_data_value) <= variable_properties["maximum_length"]
+            if not is_in_range:
+                warning_string = f"Array length more than {maximum_length}."
+                om.add_warning(warning_string, f"{var_name=}", info_map)
+
+        return is_in_range
```

Comment:
The same applies to other validation functions as well. There is no reason to obfuscate the behavior by `is_in_range`. Explanatory variables should not be used in the logic.
```suggestion
        if minimum_length is not None:
            is_in_range = variable_properties["minimum_length"] <= len(input_data_value)
            if not is_in_range:
                warning_string = f"Array length less than {minimum_length}."
                om.add_warning(warning_string, f"{var_name=}", info_map)
                return False
        if maximum_length is not None:
            is_in_range = len(input_data_value) <= variable_properties["maximum_length"]
            if not is_in_range:
                warning_string = f"Array length more than {maximum_length}."
                om.add_warning(warning_string, f"{var_name=}", info_map)
                return False

        return True
```

Reply:
Updated